locator/topology.hh

```diff
@@ -12,37 +12,128 @@
+
+    node(const node&) = default;
+    node(node&&) = delete;
+
+    const host_id& host_id() const noexcept {
+        return _host_id;
+    }
+
+    const inet_address& endpoint() const noexcept {
+        return _endpoint;
+    }
+
+    const endpoint_dc_rack& dc_rack() const noexcept {
+        return _dc_rack;
+    }
+
+    local is_local() const noexcept { return _is_local; }
+};
+
+using node_ptr = lw_shared_ptr<const node>;
+using node_set = std::unordered_set<node_ptr>;
+
 class topology {
 public:
     struct config {
+        host_id local_host_id;
+        inet_address local_endpoint;
         endpoint_dc_rack local_dc_rack;
         bool disable_proximity_sorting = false;
     };
     topology(config cfg);
-    topology(topology&&) = default;
+    topology(topology&&) noexcept;
 
     topology& operator=(topology&&) = default;
 
     future<topology> clone_gently() const;
     future<> clear_gently() noexcept;
 
+public:
+    const node_ptr& local_node() const noexcept {
+        return _local_node;
+    }
+
+    // Adds a node with given host_id, endpoint, and DC/rack.
+    node_ptr add_node(host_id id, const inet_address& ep, const endpoint_dc_rack& dr);
+
+    // Optionally updates node's current host_id, endpoint, or DC/rack.
+    // Note: the host_id may be updated from null to non-null after a new node gets a new, random host_id,
+    // or a peer node host_id may be updated when the node is replaced with another node using the same ip address.
+    node_ptr update_node(node_ptr node, std::optional<host_id> opt_id, std::optional<inet_address> opt_ep, std::optional<endpoint_dc_rack> opt_dr);
+
+    using must_exist = bool_class<struct must_exist_tag>;
+
+    // Removes a node using its host_id
+    //
+    // If host_id is not found and must_exist is true:
+    //   the function throws internal error.
+    void remove_node(host_id id, must_exist must_exist = must_exist::no);
+
+    // Finds a node by its host_id
+    //
+    // If host_id is not found:
+    //   the function throws internal error if must_exist is true,
+    //   or returns nullptr otherwise.
+    node_ptr find_node(host_id id, must_exist must_exist = must_exist::no) const noexcept;
+
+    // Finds a node by its endpoint
+    //
+    // If endpoint is not found,
+    //   the function throws internal error if must_exist is true,
+    //   or returns nullptr otherwise.
+    node_ptr find_node(const inet_address& ep, must_exist must_exist = must_exist::no) const noexcept;
+
+    // Returns true if a node with given host_id is found
+    bool has_node(host_id id) const noexcept;
+    bool has_node(inet_address id) const noexcept;
+
     /**
      * Stores current DC/rack assignment for ep
+     *
+     * Adds or updates a node with given endpoint
      */
-    void update_endpoint(const inet_address& ep, endpoint_dc_rack dr);
+    node_ptr update_endpoint(inet_address ep, std::optional<host_id> opt_id, std::optional<endpoint_dc_rack> opt_dr);
```

Comment:
Let's call it add_or_update_endpoint() then.

Reply:
will do

Reply:
renamed in v8